### PR TITLE
2966/add microsoft ca connector

### PR DESCRIPTION
**Pull request by cornelinux**

### CODE REVIEW

**Comment by plettich** [comment]
@cornelinux You should be able to calm pep8speaks down a bit by adding something like
```
    ...
    exclude: ['privacyidea/lib/caconnectors/caservice_pb2.py', 'privacyidea/lib/caconnectors/caservice_pb2_grpc.py']
    ...
```
to `.pep8speaks.yml`

**Comment by cornelinux** [comment]
> @cornelinux You should be able to calm pep8speaks down a bit by adding something like
> 
> ```
>     ...
>     exclude: ['privacyidea/lib/caconnectors/caservice_pb2.py', 'privacyidea/lib/caconnectors/caservice_pb2_grpc.py']
>     ...
> ```
> 
> to `.pep8speaks.yml`

@plettich Unfortunately looks like does not work out correctly. :-/

**Comment by plettich** [comment]
> > @cornelinux You should be able to calm pep8speaks down a bit by adding something like
> > ```
> >     ...
> >     exclude: ['privacyidea/lib/caconnectors/caservice_pb2.py', 'privacyidea/lib/caconnectors/caservice_pb2_grpc.py']
> >     ...
> > ```
> > to `.pep8speaks.yml`
> 
> @plettich Unfortunately looks like does not work out correctly. :-/

This needs to be in the master branch to take effect...
Also the correct syntax would be `exclude: ['*/caservice_pb2.py', '*/caservice_pb2_grpc.py']` (or with longer path elements)

File: privacyidea/lib/caconnectors/msca.py [hunk 1]
```diff
@@ -0,0 +1,293 @@
...
+                raise CAError("required argument '{0!s}' is missing.".format(
+                    req_key))
+
+    def set_config(self, config=None):
+        self.config = config or {}
+
+        self._check_attributes()
+        self.hostname = self.config.get(ATTR.HOSTNAME)
+        self.port = self.config.get(ATTR.PORT)
+        log.error(config)
```
Comment (plettich) [inline privacyidea/lib/caconnectors/msca.py:173]: please change this to `log.info` or something else... are we logging the configuration anyway?

File: privacyidea/lib/caconnectors/msca.py [hunk 2]
```diff
@@ -0,0 +1,293 @@
...
+        :param options: Additional options like the validity time or the
+            template or spkac=1
+        :type options: dict
+        :return: Returns the certificate object if cert was provided instantly
+        :rtype: X509 or None
+        """
+        if self.connection:
+            reply = self.connection.SubmitCR(SubmitCRRequest(cr=csr, templateName=options.get("template"),
+                                                             caName=self.ca))
+            if reply.disposition == 3:
```
Comment (plettich) [inline privacyidea/lib/caconnectors/msca.py:202]: are these dispositions defined somewhere? Should we declare some static values for this?

File: privacyidea/lib/error.py [hunk 3]
```diff
@@ -146,6 +148,17 @@ def __init__(self, description="CA error!", id=ERROR.CA):
         privacyIDEAError.__init__(self, description=description, id=id)
 
 
+class CSRError(privacyIDEAError):
```
Comment (plettich) [inline privacyidea/lib/error.py:151]: I think they should inherit from the CAError so we have one base error.

**Review by plettich — CHANGES_REQUESTED** [review]

File: privacyidea/lib/caconnectors/msca.py [hunk 4]
```diff
@@ -0,0 +1,346 @@
...
+        self.name = name
+        self.hostname = None
+        self.port = None
+        self.http_proxy = None
+        self.ca = None
+        self._connection = None
+        if config:
+            self.set_config(config)
+            self._connection = self._connect_to_worker()
+            pass
```
[the remark this answers is not in the document]
Comment (cornelinux) [reply privacyidea/lib/caconnectors/msca.py:113]: done

File: privacyidea/lib/caconnectors/msca.py [hunk 5]
```diff
@@ -0,0 +1,346 @@
...
+        """
+        Send a signing request to the Microsoft CA
+
+        options can be
+        WorkingDir: The directory where the configuration like openssl.cnf
+        can be found.
+        CSRDir: The directory, where to save the certificate signing
+        requests. This is relative to the WorkingDir.
+        CertificateDir: The directory where to save the certificates. This is
+        relative to the WorkingDir.
```
[the remark this answers is not in the document]
Comment (cornelinux) [reply privacyidea/lib/caconnectors/msca.py:189]: done

File: privacyidea/lib/caconnectors/msca.py [hunk 6]
```diff
@@ -0,0 +1,346 @@
...
+                                                                                   caName=self.ca)).cert
+                return crypto.load_certificate(crypto.FILETYPE_PEM, certificate)
+            if reply.disposition == 5:
+                log.info("cert still under submission")
+                raise CSRPending(requestId=reply.requestId)
+            else:
+                log.error("certification request could not be fullfilled! {0!s}".format(reply))
+                raise CSRError()
+
+    def view_pending_certs(self):
```
[the remark this answers is not in the document]
Comment (cornelinux) [reply privacyidea/lib/caconnectors/msca.py]: already done

File: privacyidea/lib/caconnectors/msca.py [hunk 7]
```diff
@@ -0,0 +1,346 @@
...
+        """
+        CA Manager approval
+        the token/init of would not create a certificate but a
+        pending certificate request, that needs to be approved by a
+        CA manager. So we need some kind of approve method.
+        :return:
+        """
+        pass
+
+    def request_cert(self):
```
[the remark this answers is not in the document]
Comment (cornelinux) [reply privacyidea/lib/caconnectors/msca.py]: already done

File: privacyidea/lib/tokens/certificatetoken.py [hunk 8]
```diff
@@ -458,10 +469,14 @@ def get_init_detail(self, params=None, user=None):
...
         response_detail["certificate"] = certificate
+        response_detail["rollout_state"] = self.token.rollout_state
         privatekey = self.get_tokeninfo("privatekey")
         # If there is a private key, we dump a PKCS12
         if privatekey:
-            response_detail["pkcs12"] = b64encode_and_unicode(self._create_pkcs12_bin())
+            try:
+                response_detail["pkcs12"] = b64encode_and_unicode(self._create_pkcs12_bin())
+            except Exception:
+                log.warning("Can not create PCKS12 for token {0!s}.".format(self.token.serial))
```
[the remark this answers is not in the document]
Comment (cornelinux) [reply privacyidea/lib/tokens/certificatetoken.py]: already done

File: .pep8speaks.yml [hunk 9]
```diff
@@ -31,7 +31,9 @@ pycodestyle:
         - E743 # Do not use functions named `l`, `O`, or `I`.
         - W503 # Line break before binary operator.
         - W505 # Doc line too long.
-    exclude: []
+    exclude:
+        - caservice_pb2.py
+        - caservice_pb2_grpc.py
```
[the remark this answers is not in the document]
Comment (cornelinux) [reply .pep8speaks.yml:36]: done

File: privacyidea/lib/caconnectors/msca.py [hunk 10]
```diff
@@ -0,0 +1,293 @@
...
+from subprocess import Popen, PIPE
+import yaml
+import datetime
+import shlex
+import re
+import logging
+import os
+from six import string_types
+from six.moves import input
+import traceback
```
[the remark this answers is not in the document]
Comment (cornelinux) [reply privacyidea/lib/caconnectors/msca.py:40]: done

File: privacyidea/lib/caconnectors/msca.py [hunk 11]
```diff
@@ -0,0 +1,293 @@
...
+    from privacyidea.lib.caconnectors.caservice_pb2 import (GetCAsRequest,
+                                                            GetTemplatesRequest,
+                                                            SubmitCRRequest,
+                                                            GetCertificateRequest)
+    AvailableCAConnectors.append("privacyidea.lib.caconnectors.msca.MSCAConnector")
+except ImportError:
+    log.warning("Can not import grpc modules.")
+
+from privacyidea.lib.utils import is_true
+from privacyidea.lib.error import CSRError, CSRPending
```
[the remark this answers is not in the document]
Comment (cornelinux) [reply privacyidea/lib/caconnectors/msca.py:55]: done

File: privacyidea/lib/caconnectors/msca.py [hunk 12]
```diff
@@ -0,0 +1,293 @@
...
+        """
+        config[ATTR.HTTP_PROXY] = is_true(config.get(ATTR.HTTP_PROXY))
+        return config
+
+    def _connect_to_worker(self):
+        """
+        creates a connection to the middleware and returns it.
+        """
+        channel = grpc.insecure_channel('{0!s}:{1!s}'.format(self.hostname, self.port),
+                                        options=(('grpc.enable_http_proxy', int(is_true(self.http_proxy))),))
```
[the remark this answers is not in the document]
Comment (cornelinux) [reply privacyidea/lib/caconnectors/msca.py:137]: Yes, this is not the proxy how we know it.

File: privacyidea/lib/caconnectors/msca.py [hunk 13]
```diff
@@ -0,0 +1,293 @@
...
+        try:
+            grpc.channel_ready_future(channel).result(timeout=3)
```
[the remark this answers is not in the document]
Comment (cornelinux) [reply privacyidea/lib/caconnectors/msca.py:139]: Yes, we should probably do this in the future.

File: privacyidea/lib/caconnectors/msca.py (hunk 1 above)
Comment (cornelinux) [reply privacyidea/lib/caconnectors/msca.py:173]: I remove it. This was actually some remant from some manual tests.

File: privacyidea/lib/caconnectors/msca.py (hunk 2 above)
Comment (cornelinux) [reply privacyidea/lib/caconnectors/msca.py:202]: We can do this in the future. The will also be some further return values.

File: privacyidea/lib/error.py (hunk 3 above)
Comment (cornelinux) [reply privacyidea/lib/error.py:151]: done

**Comment by cornelinux** [comment]
I have incoporated the comments into a new branch/PR which is rebased on the latest master to avoid merge conflicts.
Closing this PR.

**Pull request closed without merging**